egs/wsj/s5/steps/cleanup/internal/segment_ctm_edits_mild.py

```diff
@@ -1914,21 +1914,26 @@ def process_data(args, oov_symbol, utterance_stats, word_stats):
             if (len(split_pending_line) == 0
                     or split_pending_line[0] != cur_utterance):
                 # Read one whole utterance. Now process it.
-                (segments_for_utterance,
-                 deleted_segments_for_utterance) = get_segments_for_utterance(
-                     split_lines_of_cur_utterance, args=args,
-                     utterance_stats=utterance_stats)
-                word_stats.accumulate_for_utterance(
-                    split_lines_of_cur_utterance, segments_for_utterance)
-                write_segments_for_utterance(
-                    args.text_out, args.segments_out, cur_utterance,
-                    segments_for_utterance, oov_symbol=oov_symbol,
-                    frame_length=args.frame_length)
-                if args.ctm_edits_out is not None:
-                    print_debug_info_for_utterance(
-                        args.ctm_edits_out, split_lines_of_cur_utterance,
-                        segments_for_utterance, deleted_segments_for_utterance,
+                try:
+                    (segments_for_utterance,
+                     deleted_segments_for_utterance) = get_segments_for_utterance(
+                         split_lines_of_cur_utterance, args=args,
+                         utterance_stats=utterance_stats)
+                    word_stats.accumulate_for_utterance(
+                        split_lines_of_cur_utterance, segments_for_utterance)
+                    write_segments_for_utterance(
+                        args.text_out, args.segments_out, cur_utterance,
+                        segments_for_utterance, oov_symbol=oov_symbol,
                         frame_length=args.frame_length)
+                    if args.ctm_edits_out is not None:
+                        print_debug_info_for_utterance(
+                            args.ctm_edits_out, split_lines_of_cur_utterance,
+                            segments_for_utterance, deleted_segments_for_utterance,
+                            frame_length=args.frame_length)
+                except Exception:
+                    _global_logger.error(
```

Comment:
I am not sure if we expect it to fail but still continue. If it was only because of the previous issue, which will be fixed it is better to not make this change.

Reply:
This still fails:
```
 4 2019-05-23 04:01:47,426 [steps/cleanup/internal/segment_ctm_edits_mild.py:1397 - merge_clusters - ERROR ] Failed merging clusters [[-1], [0, 1], [2], [3], [4], [5], [6], [7], [8], [9], [10], [11], [12], [    13]]
  5 2019-05-23 04:01:47,426 [steps/cleanup/internal/segment_ctm_edits_mild.py:1947 - process_data - ERROR ] Error with utterance 190404_162901-00142500-00145500
  6 2019-05-23 04:01:47,426 [steps/cleanup/internal/segment_ctm_edits_mild.py:2060 - main - ERROR ] Failed segmenting CTM edits
  7 Traceback (most recent call last):
  8   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 2075, in <module>
  9     main()
 10   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 2041, in main
 11     word_stats=word_stats)
 12   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 1920, in process_data
 13     utterance_stats=utterance_stats)
 14   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 1551, in get_segments_for_utterance
 15     segments = merge_segments(segments, args)
 16   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 1428, in merge_segments
 17     args.max_intersegment_incorrect_words_length))
 18   File "steps/cleanup/internal/segment_ctm_edits_mild.py", line 1336, in merge_clusters
 19     (merged_segment, i, new_cluster)))
 20 TypeError: '<' not supported between instances of 'Segment' and 'Segment'
 21 # Accounting: time=1 threads=1
 22 # Finished at Thu May 23 04:01:47 EDT 2019 with status 1
```

Reply:
(it does work in python2)

Reply:
My reasoning about ignoring the problem was that I'm willing to receive one utterance/segment less results, but I want to get at least something (other utterances for which this doesn't fail), especially because I don't see a way for a user how to fix these issues.

Reply:
It seems python3 is strict about comparisons. The issue is sometimes two scores are equal and then the next element in the tuple is used for comparison. 
Line 1336-1337 can be modified instead to pick the first one arbitrarily:
```
                    heapq.heappush(heap, ((-scoring_function(merged_segment), i)
                                          (merged_segment, i, new_cluster)))
```

I believe most normal corner cases are addressed and it shouldn't fail unless there is some bug. So it is better to fail so we find out there is a bug.

Reply:
Yes, that did the job.